### PR TITLE
Add custom JSON generation

### PR DESCRIPTION
**Pull request by chrisst**
to_json will print out a more human readable representation of the api
representation. It uses api property names as json key's in order to provide
more useful breadcrumbs and viewing when collapsed.

Example output:
screen shot 2019-01-15 at 12 34 25 pm

-----------------------------------------------------------------
# [all]
no-op

### CODE REVIEW

**Comment by rileykarson** [comment]
Quick high level question: What's the use case for this change? Debugging? It isn't clear to me as a reader, and it just looks like we're adding unused code.

If this is code that's going to sit around at HEAD and only be used at debugging time, we should say so above the method in a comment and have tests to make sure we don't break it with changes like variable name changes.

**Comment by rambleraptor** [comment]
I'm gonna be a pain. Apologies for this in advance.

First, why JSON over YAML? By outputting to YAML, we gain the ability to get output YAML and then feed it immediately back into MM if necessary. We can't do that in JSON.

The other thing: For outputting we have to build some kind of .to_<yaml, json, s> methods to help with debugging. Psych has [similar things for dealing with YAML](https://stackoverflow.com/questions/10629209/how-can-i-control-which-fields-to-serialize-with-yaml) (where you get to control serialization and deserialization). If we're going to do all of these methods to handle deserialization, I think we should consider doing the same for serialization too.

That last point is a little bit of an aside though. My biggest concern is the `.to_s` method. We don't call `.to_json` anywhere (yet?). I'm scared about overriding `.to_s` and then it ends up making debugging harder down-the-line because .to_s is used in a million places throughout the Ruby compiler.

**TL;DR** why JSON and changing .to_s scares me.

**Comment by chrisst** [comment]
@rambleraptor - 
why json - specifically because we use YAML everywhere so overriding the yaml serialization is much more likely to affect actual MM functionality. If I was to use yaml I'd likely be writing the entire serializer instead of being about to shim on top of existing methods such as `to_json`. I don't really see a deserializer useful, because as you pointed out: we should use yaml if we plan on programmatically consuming this :)

`to_s` - In my experience overriding a class's to string method is pretty standard behavior and should only affect debug output. Java best practices usually mandate this. The current default `to_s` just outputs the class name and the memory address of the instance, eg: `#<Provider::Terraform::CustomCode:0x00007f87bcc4e370>` . If we have code that relies on the string serialization of a class we need to stop it, as it would be pretty bad code smell. Ruby should always be using duck typing or `is_a` for type checking.

**Comment by chrisst** [comment]
@rileykarson I was planning on adding command line options that could output this json to a file or stdout. There's already the ability to output the raw yaml, but it's pretty worthless for human consumption. However I was thinking given how overloaded the compiler.rb is right now I'll chat IRL with people about design before shimming it in.

As for tests, I can definitely add something.

**Comment by chrisst** [comment]
The background is that I'm trying to refactor the Disk properties (which are a hot mess right now) and I need a way to determine the output of my changes to a particular resource. I can output compute to a json file, apply refactors and then compare the new json output against the previously generated one. Diff tools allow me to see the difference in the two in a way that the current yaml output doesn't make easy. Happy to chat more if this doesn't make full sense.

**Comment by rambleraptor** [comment]
I'd like to get @SirGitsalot's opinion on this. I have a weak opinion against this, but it's easily changed.

**Comment by rileykarson** [comment]
Discussed over chat, but to restate my opinion here: If we need this to debug our YAML definitions, the real problem seems like we made the YAML definitions too complex. Disk/RegionDisk having 3!! inlined files (`disk_parameters.yaml`, `disk_type_properties.yaml`, and `disks.yaml`) is the real issue, and I'm for inlining all of it.

**Comment by chrisst** [comment]
I think you've conflated the problem this PR is trying to solve. The question here is "should we add a custom json serializer" not "how to best solve upstream complexity". You can still inline everything upstream with or without this addition. I added some reasons for the PR (see image of nested output) but I any tool that integrates with well formatted JSON can consume the output. 

For instance using JQ you can answer "what are all the resources supporting a kmskeyname property" using json searching syntax:
```
cat file.json | jq '[paths] | .[] | select (contains(["kmsKeyName"]))'
```

**Comment by rambleraptor** [comment]
I agree that we've drifted from the point. And I do like being able to do `jq` magicks on api.yaml

I've still got weak opinions, so I'm passing over to @SirGitsalot for a final verdict.

**Comment by chrisst** [comment]
Talking with the team today the consensus was in favor of overriding `to_s` and `to_json` to provide an easier output when using pry or trying to serialize complex objects.

**Comment by chrisst** [comment]
@rambleraptor as per conversation yesterday and in the MM meeting - the default to_s is now pretty json. This PR is ready for review
